<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
<title>IdeogramBrowser Software Documentation</title>
<link href="standard.css" rel="stylesheet" type="text/css">
<link rel="shortcut icon" href="favicon.ico" >  
</head>

<body>

<div class="back"><a href="index.html">back</a></div>


<div class="title">IdeogramBrowser Software Documentation</div>
<center><b>A Java tool for visualizing  genomic aberrations<br> using Affymetrix 
SNP arrays.</b></center>


<h1>Introduction</h1>


<table>
  <tr>
    <td>
      <br><br><br>
      
      The <span class="ideo">IdeogramBrowser</span> can be used to
      easily compare a
      set of output files from the 
      <a target="_blank" href="http://www.affymetrix.com/products/software/specific/cnat.affx">Affymetrix Chromosome Copy Number Analysis Tool</a>.
      
      Loss of heterozygosity (LOH) or loss/gain markers are plotted against their 
      corresponding chromosomal region in a karyotypic representation
      together with genes.
    </td>
    <td>
      <img alt="Chromosome" src="pics/chromosome.png">
    </td>
  </tr>
</table>



<h1>Features</h1>
IdeogramBrowser provides:
<ul>
  <li>Plotting of SNP markers in a karyogram (always new generated from
  the <tt>ideogram</tt> table of the NCBI mapview BUILD 36.2 database).
  See the <a href="#database">Appendix A</a> for changing the default database. </li>
  <li>SNP array sizes of 10k, 50k, 100k, and 500k can be imported from the Affymetrix 
  copy number analysis tool (CNAT 3 and CNAT 4) directly in the <a href="#format_B">CNT file format</a> (version 1.0 and 1.1).</li>
  <li>A general <a href="#format_A">tab delimited text file format</a> can be imported for importing foreign data
  (e.g. generated by a spreadsheet application such as Microsoft Excel). So the 
software is not limited to Affymetrix SNP arrays and the Affymetrix Copy Number Analysis Tool (CNAT 3 and 4).
  SNP analysis can be combined e.g. with array CGH data.
  The new CNAT 4 format splits the LOH markers in a second file. If the files have the same prefix the LOH files are automatically load upon opening a copy number (CN) file.</li>
  <li>Plotting of gene markers from a local copy of the
  <a target="_blank" href="ftp://ftp.ncbi.nlm.nih.gov/genomes/H_sapiens/mapview/">NCBI mapview database</a> (BUILD 36.2)</li>
  <li>Batch processing/multiple file import. Merging of split files.</li>
  <li>Thresholding of log2ratios, LOH (loss of heterozygosity), 
  SPA (single point analysis copy number), and 
  GSA (genome smoothed copy number) values for marker generation (the latter is only available for 
  chips with more than 10k SNPs).</li>
  <li>Noise reduction via filtering functionality</li>
  <li>Conensus region representation.</li>
  <li>Links to <a target="_blank" href="http://www.genecards.org">http://www.genecards.org</a></li>
  <li>Export to JPG, SVG, and PDF format.</li>
  <li>Printing functionality (the printing may not work under Linux in combination with CUPS - this is a Java 1.5/CUPS problem: export the diagrams as pdf and print e.g. with Acrobat Reader)</li>
  <li>A <em>condensed mode</em> is included showing only profile lines with aberrations. </li>
</ul>

<h1>History</h1>

<dl>
  <dt>v0.20</dt>
  <dd>
	<ul>
	  <li>Import of the CNAT 4 file format - bug fixed.</li>
	</ul>
  </dd>
  <dt>v0.19.8</dt>
  <dd>
	<ul>
	  <li>The CNAT 4 format is now supported (format version 1.1)</li>
	  <li>Different colors for amplifications and double losses.</li>
	  <li>Color swap for gains/losses.</li>
	  <li>Custom colors for aberrations.</li>
	  <li>Info panel showing the header information of a selected profile supporting marking and copying (CTRL+C) into the clipboard such that the information can easily be pasted into a text processing software.</li>
	</ul>
  </dd>
  <dt>v0.19.7</dt>
  <dd>
	<ul>
	  <li>Printing functionality for single chromosomes and all chromosomes added.</li>
	  <li>PDF export added.</li>
	  <lI>Scrolling to selected markers when selected with the left/right arrow keys.</li>
	  <li>Improved selection of overlapping regions with the mouse. By clicking more than
	  once on a marker selection all overlapping markers are cycled.</lI>
	  <li>Improved error handling for CNT file import.</li>
	  <li>Condensed mode for more compact diagrams added.</li>
	  <li>Improved zoom functionality: Click and drag within the ideogram
	  to create a zoom region.</li>
	</ul>
  </dd>
</dl>

<h1>Known Problems</h1>

<ul>
<li>There may occur problems when accessing printers from Java 1.5 under Linux using the CUPS server
(depending on the CUPS/Linux version). However this problem does not occur under WindowsXP/MacOSX.</li>
</ul>


<h1>Data Flow</h1>

<table>
  <tr>
    <td width="100%">
<ol>
  <li>Data Import of CNT or text file</li>
  <li>Thresholding for gain/loss or LOH detection</li>
  <li>Optional merging of two or more SNP profiles</li>
  <li>Pre- or post-merge filtering via a sliding window approach</li>
  <li>Consensus region finding</li>
  <li>Interactive exploration, SVG export</li>
</ol>
</td>
<td>
<img src="pics/ideo-dataflow.png" alt="IdeogramBrowser data flow">
  </td>
  </table>

<h1>Installation</h1>
The software requires the Java runtime environment 1.5.0 (JRE 5.0) which can be downloaded from <a target="_blank" href="http://java.sun.com/j2se/1.5.0/download.jsp">java.sun.com</a>.
<ol>
  <li>Install JRE 5.0 (1.5.0) or newer if it is not available on your system.</li>
  <li>Download the <a href="#download">current IdeogramBrowser version</a> and unpack the file into a directory. MacOS X users can alternatively use the .dmg disk image.</li>
  <li>Start <span class="ideo">IdeogramBrowser</span> by
  <ul>
    <li>Windows: Double click on <tt>ideogram.bat</tt> in the installation directory.</li>
    <li>Unix/Linux/MacOS X: Call <tt>ideogram.sh</tt> from the command line (ensure that your java executable is in the search path) in the installation directory.</li>
    <li>MacOS X with disk image: Open the disk image with a double click and drag the <tt>ideogram-xxx.app</tt> icon to your applications folder.</li>
  </ul>
  <b>Hint:</b> It is <b>not</b> recommended to call <span class="ideo">IdeogramBrowser</span> by a direct click on the <tt>ideogram.jar</tt> file since the standard memory assignment is too low for the most file arrangements.
</li>
</ol>
IdeogramBrowser is packed into a single zip archive which must be unzipped (e.g. using WinZip) into the desired directory (e.g. "C:\Program Files\" or your home directory). Call <tt>ideogram.sh</tt> in the IdeogramBrowser directory from the command line (Linux or Mac OS X) or double click <tt>ideogram.bat</tt> under Microsoft Windows in the file explorer.<br>
When importing many files IdeogramBrowser may report an <b>out of memory</b> error. In these cases open <tt>ideogram.sh</tt> (Linux or Mac OS X) or <tt>ideogram.bat</tt> (Windows) with a text editor (e.g. vi, Emacs, or notepad) and change the memory size by editing the two <tt>-X</tt> options of the java call. The <tt>-X</tt> options specify the memory size (in MB) assigned to the software. These can be adapted depending on the available memory and the testset size (number of imported .cnt files). The standard settings are usually too small for the most problems (<tt>2MB</tt> for Sun Java 1.5 on Linux and <tt>1MB</tt> on win). Replace the two parameters e.g. by <tt>-Xmx1024m -Xms1024m</tt> for assigning 1024MB memory to the software.


<h1><a name="download">Download</a></h1>

<h2>Pre-Compiled Version</h2>
The current IdeogramBrowser version can be downloaded from this link:
<p>
  <a href="http://www.informatik.uni-ulm.de/ni/staff/HKestler/ideo/files/IdeogramBrowser-0.20.0.zip">IdeogramBrowser-0.20.0.zip</a> (v0.20.0)
<br>

<a href="http://www.informatik.uni-ulm.de/ni/staff/HKestler/ideo/files/IdeogramBrowser-0.19.8.zip">IdeogramBrowser-0.19.8.zip</a> (v0.19.8)
<br>  
<a href="http://www.informatik.uni-ulm.de/ni/staff/HKestler/ideo/files/IdeogramBrowser-0.19.7.zip">IdeogramBrowser-0.19.7.zip</a> (v0.19.7)
<br>
<a href="http://www.informatik.uni-ulm.de/ni/staff/HKestler/ideo/files/IdeogramBrowser-0.19.5.zip">IdeogramBrowser-0.19.5.zip</a> (v0.19.5)
</p>

<h2>Source Code</h2>
<a href="http://www.informatik.uni-ulm.de/ni/staff/HKestler/ideo/files/IdeogramBrowser-0.20.0-src.tar.gz">IdeogramBrowser-0.20.0-src.tar.gz</a> (v0.20.0)<br>

<a href="http://www.informatik.uni-ulm.de/ni/staff/HKestler/ideo/files/IdeogramBrowser-0.19.8-src.tar.gz">IdeogramBrowser-0.19.8-src.tar.gz</a> (v0.19.8)<br>

<a href="http://www.informatik.uni-ulm.de/ni/staff/HKestler/ideo/files/IdeogramBrowser-0.19.7-src.tar.gz">IdeogramBrowser-0.19.7-src.tar.gz</a> (v0.19.7)

<h1>License</h1>  
<!--Creative Commons License-->
<a rel="license" href="http://creativecommons.org/licenses/by/2.5/">
<img alt="Creative Commons License" border="0" src="http://creativecommons.org/images/public/somerights20.png"></a><br>
This work is licensed under a 
<a rel="license" href="http://creativecommons.org/licenses/by/2.5/">Creative Commons Attribution 2.5  License</a>
and has to be properly cited.
<!--/Creative Commons License-->
<!-- <rdf:RDF xmlns="http://web.resource.org/cc/" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
<Work rdf:about="">
<license rdf:resource="http://creativecommons.org/licenses/by/2.5/" />
<dc:title>IdeogramBrowser</dc:title>
<dc:date>2006</dc:date>
<dc:description>Software</dc:description>
<dc:creator><Agent><dc:title>Hans A. Kestler and Andre Mueller</dc:title></Agent></dc:creator>
<dc:rights><Agent><dc:title>Hans A. Kestler and Andre Mueller</dc:title></Agent></dc:rights>
<dc:source rdf:resource="http://www.informatik.uni-ulm.de/ni/staff/HKestler/ideo/" />
</Work>
<License rdf:about="http://creativecommons.org/licenses/by/2.5/">
<permits rdf:resource="http://web.resource.org/cc/Reproduction"/>
<permits rdf:resource="http://web.resource.org/cc/Distribution"/>
<requires rdf:resource="http://web.resource.org/cc/Notice"/>
<requires rdf:resource="http://web.resource.org/cc/Attribution"/>
<permits rdf:resource="http://web.resource.org/cc/DerivativeWorks"/>
</License></rdf:RDF> -->


<a target="_blank" href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/btl608v1">Bioinformatics Abstract</a>

<h1>Included Libraries</h1>

IdeogramBrowser includes the following open source Java libraries:
<ul>
<li>JUnit v4.1<br>
<a target="_blank" href="http://www.junit.org">http://www.junit.org</a>
</li>

<li>XStream v1.2<br>
<a target="_blank" href="http://xstream.codehaus.org/">http://xstream.codehaus.org/</a>
</li>

<li>Batik v1.6<br>
<a target="_blank" href="http://xmlgraphics.apache.org/batik/">http://xmlgraphics.apache.org/batik/</a>
</li>

<li>IText v1.4.6<br>
<a target="_blank" href="http://www.lowagie.com/iText/">http://www.lowagie.com/iText/</a>
</li>
</ul>


<h1><a name="ideo_main">Running IdeogramBrowser</a></h1>

The following steps lead to a chromosomal displayed marker set:
<ol>
  <li><a href="#import">Import a marker set</a></li>
  <li><a href="#filtering">Filtering</a> (Choose copy number value and threshold)</li>
  <li><a href="#explore">Explore the displayed ideogram</a></li>
</ol>

<h2><a name="format">Import File Formats</a></h2>
IdeogramBrowser can import two different file formats:

<h3><a name="format_A">Tabbed Text File Format</a></h3>
This is a tab delimited ASCII file format with one aberration value in each line which can easily be generated by e.g. Microsoft Excel or OpenOffice.
<p>
Each marker line should contain five columns:<br>
<tt>
CHROMOSOME (1..22,X,Y)<br>
START_BP (Integer &gt; 0)<br>
STOP_BP (Integer &gt; 0)<br>
NAME (String)<br>
VALUE (Integer)<br>
</tt>
A short example can be downloaded: <a href="files/marker_example.txt">marker_example.txt</a>.
The <tt>VALUE</tt> field represents a copy number which is defined by the following ranges:
<table>
  <tr><td>VALUE &lt; 0</td><td>Illegal value</td></tr>
  <tr><td>0 &lt; VALUE &lt; 2</td><td>Loss</td></tr>
  <tr><td>2</td><td>No aberration</td></tr>
  <tr><td>VALUE &gt; 2</td><td>Gain</td></tr>  
</table>
</p>

<h3><a name="format_B">CNAT Export File Format</a></h3>
IdeogramBrowser can import the 
<a target="_blank" href="http://www.affymetrix.com/products/software/specific/cnat.affx">Affymetrix Chromosome Copy Number Analyis Tool (CNAT)</a> 
output file format (.cnt files). 
The tab-delimited TXT format of the CNAT tool is currently not supported.
A brief summary of using CNAT to generate these files is given in the CNAT manual. 
Example files can be downloaded: <a target="_blank" href="http://www.affymetrix.com/support/technical/sample_data/100k_copynumber_data.affx">Affymetrix sample data sets</a> (these have to be first processed by CNAT 3.0 or newer to compute
the copy numbers).
The Affymetrix example data sets MCF7, Ref103, SKBR3, and ZR75-30 were processed by CNAT 3.0 with
the standard settings and the corresponding CNT files can be downloaded <a href="http://www.informatik.uni-ulm.de/ni/staff/HKestler/ideo/files/ideo-examples.zip">ideo-examples.zip</a>


<h2><a name="import">Importing Input Files</a></h2>
<h3>Tabbed Text File</h3>
Choose the menu point <em>Load Markers</em> in the <em>File</em> menu and navigate to a file in the format described in <a href="#format_A">text file format</a>.

<h3>CNAT Export File</h3>
Choose the menu point <em>Load Markers</em> in the <em>File</em> menu and navigate to a file in the format described in <a href="#format_A">CNT file format</a>.


<h2><a name="filtering">Filtering</a></h2>
The imported marker sets can be thresholded and filtered.  All required
parameters can be controlled via the filtering panel. Upon pressing
the <em>OK</em> button the visualization is immediately adapted.
<center><img alt="IdeogramBrowser filtering panel"
src="pics/filter_panel.png"></center>


The following parameters can be adapted:
<dl>
  <dt>Colors</dt>
  <dd>With clicking on the <em>inv</em> button the colors for
	gains/losses are switched.  A click with the right mouse button on
	a color shows a color selector from which a customized color can
	be selected.
  </dd>
  
  <dt>Display Modes</dt>
  <dd>
	<em>Condensed Mode</em> Shows only non-empty marker lines if activated.<br>
	<em>Show Lines</em> Shows a gray background line for each sample
  </dd>
  <dt>Field Selection</dt>
  <dd>
	Depending on the loaded CNAT format version (3 or 4) different fields appear in the field selector.
    Using the drop down menu on the
    filter panel switches between the display of
	<dl>
	  <dt>CNAT 3 format</dt>
	  <dd><em>LOGRATIO</em>, <em>GSA</em>, <em>SPA</em>, <em>LOH</em></dd>
	  <dt>CNAT 4 format</dt>
	  <dd><em>CNState, Log2Ratio, HmmMedianLog2Ratio, NegLog10Pvalue, LOHProb</em></dd>
	  </dl>
		values from the CNT  files. In case of tabbed text file input only copy numbers are
    supported and displayed (see <a href="#format_A">tabbed text file format</a>).
  </dd>

    
  <dt>Thresholding</dt>
  <dd>The imported values could be thresholded by tuning the
    <em>lower</em> and <em>upper</em> bounds for the copy number detection.  After
    pressing the OK-Button all markers with values lower/greater than the
    lower/upper bound will be displayed in the ideogram on the left/right
    side of the chromosomes.
  </dd>

  <dt>Merging Method</dt>
  <dd>
    When merging multiple SNP files into a single profile via the
    context menu <em>Merge Marker File</em> among two different
    alternatives can be chosen:
    <ul>
      <li><em>resolution enhancement:</em>Interlocking SNP chips are
      merged in order to extend the genomic resolution.</li>
      <li><em>reliability enhancement/conflict detection:</em>Interlocking SNP chips or
      SNP chips of the same type are scanned for common gains/losses
      which are then shown in the diagram. If one chip votes for a
      gain and the other chip(s) votes for a loss this conflict
      situation is marked by a <b>magenta</b> marker.
      This mode can be activated by setting the check box
      <em>Diff. Mode</em> in the filter panel.
      </li>
      
    </ul>
  </dd>
  
  <dt>Filtering Method</dt>
  <dd>
    It can be choosen among three filtering routings:
    <ul>
      <li><em>no filtering:</em> All SNPs are shown</li>
      <li><em>pre-filtering:</em> SNPs are filtered before merging</li>
      <li><em>post-filtering:</em> SNPs are filtered after merging</li>
    </ul>
    For single profiles the last two methods are identically.
  </dd>

  <dt>Filtering Parameters</dt>
  <dd>Group limit is used to restrict the view on
    groups of adjacent markers with the given minimal group
    size. Additionally a minimal length for the length of this group in
    million base pairs could be set. Both criteria have to be fulfilled
    for a certain region in order to be shown in the diagram.</dd>

  <dt>Consensus Mode</dt>
  <dd>If the check box <em>Consensus Mode</em> is activated a further
    marker line is shown in the diagram. Only those regions are shown
    for which at least <em>n</em> number of samples vote for gain/loss
    or LOH. This limit is configurable via the spin button on the
    right side of the consensus mode check box.
  </dd>
</dl>

<h2><a name="explore">Interactive Exploration</a></h2>
IdeogramBrowser allows the interactive exploration of the marker sets.
<center><!-- <a href="ideogram_panel_big.html"><img src="pics/ideogram_panel.png"></a>--> 
<img src="pics/ideo_screen.png" alt="IdeogramBrowser screenshot">
</center>
<p>
The upper (ideogram) panel shows the 24 chromosomes and the imported markers. When moving the mouse slowly over the chromosomes, the corresponding genome region under the mouse is shown in a small tool tip window. Also the GeneID and gene names are shown when moving the mouse over the blue bar on the right side of the chromosomes. <br>
When a gene is selected with the left mouse button the details of the selection are listed in the info panel below the ideogram panel. If a marker is selected (blinking) the marker info and a summary of intersecting genes is displayed on the info panel.
</p>

<h3>The Karyogram Panel</h3>
<p>
Depending how many datasets are loaded into IdeogramBrowser one marker
line appears on the left and the right side of the ideogram. A data
slot can be selected with the mouse by clicking on the gray marker
line. Once a data slot is selected all SNPs (independent of the merge
mode and the number of losses/gains or LOHs) are shown on the left
side for an assessment of the coverage of the genome.
</p>
<center><img alt="karyogram panel" src="pics/ideo_chromosome.png"></center>

<h3>The Info Panel</h3>
<p>
The lower (info) panel shows the information for the selected items on the ideogram panel.
Marker info and/or gene info is available. Clicking on the hyperlinked gene names opens a 
browser window displaying additional information from the 
<a href="http://www.genecards.org" target="parent">www.genecards.org</a> webpage.
</p>
<center><img src="pics/info_panel.png" alt="IdeogramBrowser info panel"></center>

<h3><a name="tooltip">Tool Tip Info Window</a></h3>
When moving the mouse over chromosomes, genes, or markers a small tool tip window
appears showing information about the chromosomal region, the genes or
the SNP markers.

<h3><a name="popupmenu">Popup Menu</a></h3>
<table>
  <tr><td>
    By clicking the right mouse button on a marker/marker line 
    the popup menu opens. 
    It provides <em>Zoom In</em> and <em>Zoom Out</em> fields for zooming into the chromosomes.
	<em>Show All</em> shows the whole chromosome.
    If a marker is selected additional the <em>Remove Marker File</em> option is shown. 
    By choosing this command the complete marker line containing the selected marker 
    is removed from the workbench.
    The context menu item <em>Merge marker file</em> enables to merge the 
    currently selected marker line with another CNT file. This feature is useful for
    importing split files such as those originating from the CNAT analysis
    of 100k chips.
  </td>
  <td width="0"><center><img src="pics/popup_menu.png" alt="IdeogramBrowser context menu"></center></td></tr>
</table>
  
  <h3><a name="keys">Keyboard and Mouse Commands</a></h3>
  The ideogram panel could be controlled by clicking the folowing keys:
  <ul>
    <li>Select one chromosome with the mouse.</li>
    <li>Press <tt>+/-</tt> to <em>zoom</em> in/out the chromosome.</li>
    <li>Press <tt>Up, Down, Page Up, Page Down, Home or End</tt> or move mouse wheel to <em>scroll</em> through a zoomed chromosome.</li>
    <li>Press <tt>Shift</tt> and move the mouse wheel simultaneously
    to <em>zoom</em> in and out.
    <li>Press <tt>Enter or Second Mouse Button</tt> to <em>reset the zoom level</em> of the chromosome.</li>
    <li>Press <tt>Right/Left</tt> to step to the <em>next/previous marker</em>. This command is only available if a marker is selected.</li>
	<li>Click into the ideogram and drag to select a zooming window which is then accordingly selected. Double clicking or using the <em>Zoom Out</em> function restores the last zoom setting.</li>
  </ul>


  <h2><a name="save">Saving a Karyogram</a></h2>
  Select <em>File/Save</em> to save the current karyogram as JPEG, PDF, or SVG (scalable vector graphics).
  <b>Hint:</b> SVG export may take a very long time.
  <br>
  <br>
  <p>The original version of this page is available at
  <a href="http://www.informatik.uni-ulm.de/ni/staff/HKestler/ideo/doc.html">http://www.informatik.uni-ulm.de/ni/staff/HKestler/ideo/doc.html</a>
  </p>

  <h2><a name="print">Printing</a></h2>
  Single chromosomes or the whole ideogram can be printed via the two menu items
  <ul>
	<li><em>File/Print single chromosome</em></li>
	<li><em>File/Print all chromosomes</em></li>
  </ul>
 <b>Hint:</b>The printing may not work under Linux in combination with CUPS -
  this is a Java 1.5/CUPS problem: export the diagrams as pdf and print e.g. with Acrobat reader.

  <h1><a name="appendix">Appendix</a></h1>

  <h3><a name="appendix_a">The gene/ideogram database</a></h3>

  
  The two database table <tt>ideogram.gz</tt> and <tt>seq_gene.md.gz</tt> are packed into the <em>jar</em> archive of IdeogramBrowser. The current version of this database are available on the NCBI ftp site:
  <a target="_blank" href="ftp://ftp.ncbi.nlm.nih.gov/genomes/H_sapiens/mapview/">ftp://ftp.ncbi.nlm.nih.gov/genomes/H_sapiens/mapview/</a>.
  Older versions can be found in the archives
  <a target="_blank" href="ftp://ftp.ncbi.nlm.nih.gov/genomes/H_sapiens/ARCHIVE">ftp://ftp.ncbi.nlm.nih.gov/genomes/H_sapiens/ARCHIVE</a>. Just place the above mentioned files into the <tt>ideogram/data</tt> directory in the archive (unjar and jar the archive with the <tt>manifest.txt</tt> file).
  As the file format of these tables changes for (nearly) every release it may be necessary to
  <ol>
	<li>adapt the <tt>read()</tt> functions in <tt>ideogram.db.GeneDB</tt> and <tt>ideogram.db.IdeogramDB</tt>, or</li>
	<li>reformat the database files such that they have the BUILD.36.2 format.</li>
  </ol>
  
  <hr>
  <div class="back"><a href="index.html">back</a></div>
  <address>
  <a href="mailto:hans.kestler@uni-ulm.de">Hans A. Kestler</a>
  </address>
  <div class="modified">
  &copy; by Hans A. Kestler &amp; Andr&eacute; M&uuml;ller<br>
  <!-- hhmts start -->
<!-- hhmts end -->
  </div>
  </body>
  </html>

